Implement support for multiple vault connections on one project

Hakon opened the pull request:
This implements support for multiple namespaced vault connections. This should solve issue #10 as well.
This is a breaking change in the sense that the env parameter for the vault token has changed. I can implement backwards compatibility in the cases where only one connection is specified.

What do you think?

Commit:
Implement support for multiple vault connections on one project

VladRassokhin commented on server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultBuildStartContextProcessor.kt:
```diff
@@ -31,39 +31,45 @@ class VaultBuildStartContextProcessor(private val connector: VaultConnector) : B
+            val buildType = build.buildType ?: return emptyList()
 
-            val connectionFeature = buildType.project.getAvailableFeaturesOfType(OAuthConstants.FEATURE_TYPE).firstOrNull {
+            val connectionFeatures = buildType.project.getAvailableFeaturesOfType(OAuthConstants.FEATURE_TYPE).filter {
```
Let's carefully process all returned features from `getAvailableFeaturesOfType` method. 
1. There's possibility that feature with same prefix could exist in project and it's parent project. Some test is required for that.
2. Now you could have two project features with same prefix. It should be processed somehow and probably some error shown.

Hakon replied:
You can also just make two features on the same project, I haven't implemented any uniqueness-check. We can check if there are multiple features with the same prefix and log an error in case there are.

VladRassokhin replied:
I've resolved that and pushed commit into your master branch

VladRassokhin commented:
Thank you for contribution, I'd try it out tomorrow and probably would comment something else.

Hakon commented:
I have had some thoughts about the naming of "prefix", is "namespace" a better name?

Commit:
Add build problem if multiple vault connections are defined with same prefix

Hakon commented:
As it stands now, the plugin postfixes the configuration parameters with the vault prefix (counter intuitive, i know) like this:

```teamcity.vault.wrapped.token.test-vault```

Is it better to have the prefix actually prefix the rest of the parameter name?

```teamcity.test-vault.wrapped.token```

In the case where you only have 1 connection called vault the parameter name is the same as before.

VladRassokhin commented:
Maybe something like `teamcity.vault.test.wrapped.token` for custom and for default `teamcity.vault.default.wrapped.token` or even just `teamcity.vault.wrapped.token`?

VladRassokhin committed:
Improve several features support:

Two features with same prefix cannot coexist in same project
Though it's possible to override feature with same prefix in subproject

VladRassokhin commented:
@Hakon Could you please sign CLA if you haven't yet? https://www.jetbrains.com/agreements/cla/

VladRassokhin commented on agent/src/main/kotlin/org/jetbrains/teamcity/vault/agent/VaultBuildFeature.kt:
```diff
@@ -53,32 +53,38 @@ class VaultBuildFeature(dispatcher: EventDispatcher<AgentLifeCycleListener>,
+            if(url == null || url.isNullOrBlank())
+                return@forEach
+            val logger = runningBuild.buildLogger
+            logger.activity("HashiCorp \"$prefix\" Vault", VaultConstants.FeatureSettings.FEATURE_TYPE) {
```
Let's not split tool name, `HashiCorp Vault ($prefix)` is better

VladRassokhin commented on common/src/main/kotlin/org/jetbrains/teamcity/vault/VaultFeatureSettings.kt:
```diff
@@ -15,11 +15,16 @@
 package org.jetbrains.teamcity.vault
 
-data class VaultFeatureSettings(val url: String, val endpoint: String, val roleId: String, val secretId: String) {
+data class VaultFeatureSettings(val parameterPrefix: String, val url: String, val endpoint: String, val roleId: String, val secretId: String) {
```
Probably could be simplified to `prefix`.
Though I've idea to restrict possible prefixes like: `vault:`,`vault-X:`, `vault-Y:` and so. It would be easier to distinguish such parameters by human.

Hakon replied:
Then maybe I should rename the prefix to namespace

Hakon replied:
I couldn't find the "reformat changed code" option (i am a bit new to intellij). I did a regular interactive rebase and reformatted the parts that were in conflict with "reformat code"

VladRassokhin reviewed (dismissed):
Also could you please *rebase* your branch onto current master. And while interactively rebasing use 'reformat changed code' in IntelliJ. Thanks!

Hakon commented:
CLA has been signed.
I could not get the project to load in intellij, there were some modules missing. 
```diff --git a/.idea/modules.xml b/.idea/modules.xml
index 889720b..bd78d19 100644
--- a/.idea/modules.xml
+++ b/.idea/modules.xml
@@ -2,16 +2,12 @@
 <project version="4">
   <component name="ProjectModuleManager">
     <modules>
-      <module fileurl="file://$PROJECT_DIR$/.idea/modules/agent/agent.iml" filepath="$PROJECT_DIR$/.idea/modules/agent/agent.iml" group="agent" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/agent/agent_main.iml" filepath="$PROJECT_DIR$/.idea/modules/agent/agent_main.iml" group="agent" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/agent/agent_test.iml" filepath="$PROJECT_DIR$/.idea/modules/agent/agent_test.iml" group="agent" />
-      <module fileurl="file://$PROJECT_DIR$/.idea/modules/common/common.iml" filepath="$PROJECT_DIR$/.idea/modules/common/common.iml" group="common" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/common/common_main.iml" filepath="$PROJECT_DIR$/.idea/modules/common/common_main.iml" group="common" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/common/common_test.iml" filepath="$PROJECT_DIR$/.idea/modules/common/common_test.iml" group="common" />
-      <module fileurl="file://$PROJECT_DIR$/.idea/modules/server/server.iml" filepath="$PROJECT_DIR$/.idea/modules/server/server.iml" group="server" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/server/server_main.iml" filepath="$PROJECT_DIR$/.idea/modules/server/server_main.iml" group="server" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/server/server_test.iml" filepath="$PROJECT_DIR$/.idea/modules/server/server_test.iml" group="server" />
-      <module fileurl="file://$PROJECT_DIR$/.idea/modules/teamcity-hashicorp-vault-plugin.iml" filepath="$PROJECT_DIR$/.idea/modules/teamcity-hashicorp-vault-plugin.iml" />
     </modules>
   </component>
 </project>
\ No newline at end of file```

Commit:
Don't split tool name in log

Hakon commented:
Something like ```teamcity.vault.default.wrapped.token``` would break existing installs, is this problematic?

VladRassokhin commented:
Just ignore unloaded modules or try reimporting project from build.gradle

>Something like `teamcity.vault.default.wrapped.token` would break existing installs, is this problematic?

That's why I'd prefer `teamcity.vault.wrapped.token`. AFAIU that's doable

Hakon commented:
I agree with teamcity.vault.<prefix>.wrapped.token or teamcity.vault.wrapped.token if default.

I also plan to make the following changes:
VAULT_TOKEN.test-vault will be renamed to <PREFIX>_VAULT_TOKEN, and VAULT_TOKEN if default
Same goes for VAULT_ADDR.

I realized that VAULT_TOKEN.test-vault is not a valid environment variable in for example bash.

Commit:
Redo namespacing of parameters and environment variables

Commit:
Enable each vault connection independently.
This adds the configuration parameters teamcity.vault.set.env.<prefix>

Fix case where no vault integrations were activated.
This happened when the first vault connection was not used, the VaultParametersProvider would return early without setting up the rest of the integrations.

Hakon commented:
I changed so that the vault connections are now enabled independently so if a build configuration has need for one connection none of the others needs to fetch token. With this I also added the ```teamcity.vault.set.env.<prefix>``` configuration variable to allow for environment variables from a subset of the connections.

VladRassokhin committed:
Reuse code when working with prefix-related properties

VladRassokhin committed:
Use 'namespace' approach.

Possible references are: `%vault:/path%`, `%vault:namespace:/path%`

VladRassokhin commented:
Hi @Hakon.
Thank you for contribution, I've finally had time to review and test it.
I've changed name to 'namespace', changed how references should be constructed, so now there's always `vault:` prefix.
